# Reliability of the borrowers

## Let's open the dataset and examine the general information about the data.

**Task 1. We are going to import the `pandas` library. Read the data from the .csv file into a dataframe and store it in the `data` variable. The path to the file:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('data.csv')

pd.options.mode.chained_assignment = None

**Task 2. The first 20 lines of the `data` dataframe on the screen.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Task 3. Print basic information about the dataframe using the method `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Data preprocessing

### Removing missing values

**Task 4. Using a combination of the two methods let's print the number of missing values for each column.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Task 5. There are missing values in two columns. One of them is `days_employed`. Missing values in that column will be processed in the next step. Another column with missing values is `total_income`, which stores person's income data. The amount of income is most affected by the type of employment, so let's fill in the missing values in this column with the median value for each type from the `income_type` column. For example, for a person with an employment type of `employee`, the gap in the `total_income` column will be filled by the median income among all records of the same type.**

In [5]:
# iterate through each employment type in the set of unique values of the income_type column
# at each step of the loop, using the loc attribute we select rows,
# where income_type is the current type of employment (d) and there are missing values in total_income
# and write the average values of total_income among the rows with the current type of device (d)
# check that all gaps are filled

for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Handling data anomalies (artifacts)

**Task 6. Data artifacts (anomalies) can occur inside the data - values that do not reflect reality and appear due to some kind of error. In our case these artifacts would be the negative number of days of employment in the `days_employed` column. For raw data, this is normal. We will handle the values in this column: replacing all negative values with positive ones using the method `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Task 7. For each type of employment, print the median value of `days_employed` in days.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

Two types (unemployed and retired persons) will get abnormally large values. Correcting such values is difficult, so we leave them as they are. Moreover, we will not need this column for our research.

**Task 8. List the unique values of the `children` column.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Task 9. There are two anomalies in the `children` column. Let's remove rows containing such anomalous values from the dataframe `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Task 10. Next, we list the unique values of the `children` column again to make sure the artifacts have been removed.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Removing missing values (continued)

**Task 11. We fill in the missing values in the `days_employed` column with the median values for each `income_type` employment type.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Task 12. Let's make sure all gaps are filled in. Test yourself and print the number of missing values for each column again using two methods.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Data Type Conversion

**Task 13. Now, we replace the real data type in the `total_income` column with an integer using the `astype()` method.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Handling Duplicates

**Task 14. Let's now handle implicit duplicates in the `education` column. This column has the same values, but written slightly differently (e.g. using uppercase and lowercase letters). Let's onvert them to lower case. Check the rest of the columns.**

In [14]:
data['education'] = data['education'].str.lower()

**Task 15. Next step is to display the number of duplicate rows in the data. If such lines are present, remove them.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Data classification / categorization

**Task 16. Based on the ranges below, let's create a `total_income_category` column in the `data` dataframe with the following categories:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**For example, a potential borrower with an income of 25,000 should be assigned an `'E'` category, and a customer with an income of 235,000 should be assigned a `'B'` category. Use your own function named `categorize_income()` and method `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Task 17. Next step is to display a list of unique loan purposes from the `purpose` column.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Task 18. We are going to create a function which, based on the data in the `purpose` column, will generate a new `purpose_category` column, which will include the following categories:**

- `transactions with cars`,
- `real estate transactions`,
- `wedding`,
- `education purposes`.

**For example, if the `purpose` column contains the substring `'car purchase'`, then the `purpose_category` column should contain the string `'car operations'`.**

**We use our own function named `categorize_purpose()` and method `apply()`. Examine the data in the `purpose` column and determine which substrings will help you correctly identify the category.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)
data

children  days_employed  dob_years education  education_id  \
0             1    8437.673028         42    высшее             0   
1             1    4024.803754         36   среднее             1   
2             0    5623.422610         33   среднее             1   
3             3    4124.747207         32   среднее             1   
4             0  340266.072047         53   среднее             1   
...         ...            ...        ...       ...           ...   
21520         1    4529.316663         43   среднее             1   
21521         0  343937.404131         67   среднее             1   
21522         1    2113.346888         38   среднее             1   
21523         3    3112.481705         38   среднее             1   
21524         2    1984.507589         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

       total_income                       purpose total_income_category  \
0            253875                 покупка жилья                     B   
1            112080       приобретение автомобиля                     C   
2            145885                 покупка жилья                     C   
3            267628    дополнительное образование                     B   
4            158616               сыграть свадьбу                     C   
...             ...                           ...                   ...   
21520        224791             операции с жильем                     B   
21521        155999          сделка с автомобилем                     C   
21522         89672                  недвижимость                     C   
21523        244093  на покупку своего автомобиля                     B   
21524         82047         на покупку автомобиля                     C   

               purpose_category  
0      операции с недвижимостью  
1        операции с автомобилем  
2      операции с недвижимостью  
3         получение образования  
4            проведение свадьбы  
...                         ...  
21520  операции с недвижимостью  
21521    операции с автомобилем  
21522  операции с недвижимостью  
21523    операции с автомобилем  
21524    операции с автомобилем  

[21331 rows x 14 columns]

### Step 3. Explore the Data and answer the questions

#### 3.1 Is there a relationship between the number of children and timely loan repayment?

In [22]:
# define a function for calculating the percentage, which will make the conclusion in our study more understandable 
def percentage(pdSerises):
    return str(round((pdSerises.sum() / pdSerises.count()) * 100, 2)) + '%'

# to summarize the data and visualize it, we will build a pivot table
data_pivot = data.pivot_table(index=['children'], values=["debt"], aggfunc=['sum', 'count', percentage])

# sort values by number of children to get an ordered picture
data_pivot = data_pivot.sort_values(by=('children'), ascending=True)
data_pivot

sum  count percentage
          debt   debt       debt
children                        
0         1063  14091      7.54%
1          444   4808      9.23%
2          194   2052      9.45%
3           27    330      8.18%
4            4     41      9.76%
5            0      9       0.0%

**Conclusion:**

After analyzing the pivot table, we see that the presence of children in the borrower increases the likelihood of debt repayment on time, with a difference of 1.69% in case when client doesn't have any children. It might be assumed that in the event of a difficult financial situation, the client would prefer to delay the payment to the bank, rather than not provide their children with necessary things.

It is interesting to note that for families with five children, the delinquency rate seems to be zero. Based on the fact that for large children with four children, the share of delinquencies is maximum, it would be erroneous to conclude that families with many children with five children are more reliable borrowers than families with four children. This category can be characterized as "noise" in the data (or data outliers), and, accordingly, we will not draw conclusions about the reliability of borrowers based on this information.

More detailed conclusions would be possible by calculating the median for the amount of loans requested, but these data are not in our dataset.

Outcome:
- the least reliable borrowers, with the highest share of loan delinquency of 9.76%, are families with many children;
- the most reliable borrowers, with the lowest share of loan delinquency of 7.54% - are borrowers without children.

#### 3.2 Is there a relationship between marital status and loan repayment on time?

In [23]:
# to summarize the data and visualize it, we will build a pivot table
data_pivot = data.pivot_table(index=['family_status'], values=["debt"], aggfunc=['sum', 'count', percentage])

# sort values by number of children to get an ordered picture
data_pivot = data_pivot.sort_values(by=('percentage', 'debt'), ascending=False)
data_pivot

sum  count percentage
                      debt   debt       debt
family_status                               
Не женат / не замужем  273   2796      9.76%
гражданский брак       385   4134      9.31%
женат / замужем        927  12261      7.56%
в разводе               84   1189      7.06%
вдовец / вдова          63    951      6.62%

**Conclusion:**

Based on our results, if the client has ever been married (categories `married`, `divorced`, `widower/widow`), then the probability of debt will be 2-3% lower than that of a borrower who does not is officially married. Married people must make all decisions jointly. This also applies to bank loans. Usually, both people work in a marriage, so it is much easier to fulfill their obligations to a financial institution. It can be assumed that family relationships teach responsibility, including financial responsibility.

Despite the fact that in paragraph 2.7.2 it was concluded that families without children have the least debt, now it can be assumed that in order to be a reliable borrower for a bank, there is the influence of several criteria at once - the absence of children and official marriage.

The smallest share of loan delinquencies (6.62%) was found among widowed spouses.

#### 3.3 Is there a relationship between income and loan repayment on time?

In [24]:
# to summarize the data and visualize it, we will build a pivot table
data_pivot = data.pivot_table(index=['total_income_category'], values=["debt"], aggfunc=['sum', 'count', percentage])

# sort values by number of children to get an ordered picture
data_pivot = data_pivot.sort_values(by=('percentage', 'debt'), ascending=False)
data_pivot

sum  count percentage
                       debt   debt       debt
total_income_category                        
E                         2     22      9.09%
C                      1353  15921       8.5%
A                         2     25       8.0%
B                       354   5014      7.06%
D                        21    349      6.02%

**Conclusion:**

Based on the data obtained from the pivot table, we notice that inside income categories `A`, `D`, `E` the amount of data is much less than in other categories. Thus, we are faced with a data feature called data outliers or noise in the data, which, for example, could be caused by initial missing_values in the total_income column. Therefore, data in income categories `A`, `D`, `E` will not be an indicator of the reliability of a potential borrower in this analysis. To conduct a reliable analysis, it will be necessary to use quantiles.

Let's determine the dependence based on the results for categories `B`, `C`. The percentage of overdue loans is slightly lower for category `B` borrowers with an income of 200,001–1,000,000 rubles (7.06% default), compared to category `C`, in which the loan default rate is 8.5%. Based on this, we can deduce the dependence, the higher the level of income, the lower the probability of delay. At the same time, the output by categories `B` and `C` can be extended to all groups, since in total 98% of clients are in groups `B` and `C`.

#### 3.4 How do different purposes of a loan affect its repayment on time?

In [25]:
# to summarize the data and visualize it, we will build a pivot table
data_pivot = data.pivot_table(index=['purpose_category'], values=["debt"], aggfunc=['sum', 'count', percentage])

# sort values by number of children to get an ordered picture
data_pivot = data_pivot.sort_values(by=('percentage', 'debt'), ascending=False)
data_pivot

sum  count percentage
                         debt   debt       debt
purpose_category                               
операции с автомобилем    400   4279      9.35%
получение образования     369   3988      9.25%
проведение свадьбы        183   2313      7.91%
операции с недвижимостью  780  10751      7.26%

**Conclusion:**

Our pivot table shows, that there are no categories that differ greatly in the number of categories, so the result can be interpreted as follows:
- Borrowers whose purpose of the loan is `car operations` (delay rate 9.35%) or `education` (delay rate 9.25%) are more likely to default on the loan on time.
- categories of clients who take loans for `wedding` (percentage of overdue 7.91%) and `real estate` (percentage of delinquency 7.26%) are less likely to default on the loan on time.

#### 3.5 State a couple of possible reasons of missing values in the original data.

*Answer:* Common reasons leading to missing data may be: the impossibility of obtaining data or processing it; misrepresentation or concealment of information; fraud; technical reasons.

#### 3.6 Explain why filling missing data with the median is the best solution for quantitive variables.

*Answer:* In the case when some data values stand out strongly from other values, the use of an average value may incorrectly characterize the data. The median does not have the disadvantages of the mean, as it is not affected by outliers or "noise" in the data.

### Step 4. General Conclusion

The relationship between the number of children and the marital status of the client was identified after a thorough study of the data.The maximum percentage of loan nonpayments could be noticed with clients who were not officially married, as well as for clients with children. Clients included in these categories are clients with the highest risk of not repaying the loan in due time.

We also see that the higher the level of income,the less likely it is to delay the repayment of the loan.Although the peculiarity of the data in the income category of borrowers did notallow us to give an exhaustive answer for all categories,the conclusion for categories `B` and `C` can be extended to all other categories, since in total in groups B and C about 98% clients.
    
Based on our analysis, we can design the profile of the most solvent customer as follows:
    
*A person who is married, without children, with a high income, and whose purpose of obtaining a loan is real estate transactions.*